<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML0101ENSkillsNetwork20718538-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo">
    </a>
</p>

<h1 align="center"><font size="5">Final Project: House Sales in King County, USA </font></h1>


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
    <li><a href="#Instructions">Instructions</a></li>
    <li><a href="#About-the-Dataset">About the Dataset</a></li>
    <li><a href="#Module-1:-Importing-Data-Sets">Module 1: Importing Data </a></li>
    <li><a href="#Module-2:-Data-Wrangling">Module 2: Data Wrangling</a> </li>
    <li><a href="#Module-3:-Exploratory-Data-Analysis">Module 3: Exploratory Data Analysis</a></li>
    <li><a href="#Module-4:-Model-Development">Module 4: Model Development</a></li>
    <li><a href="#Module-5:-Model-Evaluation-and-Refinement">Module 5: Model Evaluation and Refinement</a></li>
</a></li>
</div>
<p>Estimated Time Needed: <strong>75 min</strong></p>
</div>

<hr>


# Instructions


In this assignment, you are a Data Analyst working at a Real Estate Investment Trust. The Trust would like to start investing in Residential real estate. You are tasked with determining the market price of a house given a set of features. You will analyze and predict housing prices using attributes or features such as square footage, number of bedrooms, number of floors, and so on. This is a template notebook; your job is to complete the ten questions. Some hints to the questions are given.

As you are completing this notebook, take and save the **screenshots** of the final outputs of your solutions (e.g., final charts, tables, calculation results etc.). They will need to be shared in the following Peer Review section of the Final Project module.


# About the Dataset

This dataset contains house sale prices for King County, which includes Seattle. It includes homes sold between May 2014 and May 2015. It was taken from [here](https://www.kaggle.com/harlfoxem/housesalesprediction?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-wwwcourseraorg-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2022-01-01). It was also slightly modified for the purposes of this course. 


| Variable      | Description                                                                                                 |
| ------------- | ----------------------------------------------------------------------------------------------------------- |
| id            | A notation for a house                                                                                      |
| date          | Date house was sold                                                                                         |
| price         | Price is prediction target                                                                                  |
| bedrooms      | Number of bedrooms                                                                                          |
| bathrooms     | Number of bathrooms                                                                                         |
| sqft_living   | Square footage of the home                                                                                  |
| sqft_lot      | Square footage of the lot                                                                                   |
| floors        | Total floors (levels) in house                                                                              |
| waterfront    | House which has a view to a waterfront                                                                      |
| view          | Has been viewed                                                                                             |
| condition     | How good the condition is overall                                                                           |
| grade         | overall grade given to the housing unit, based on King County grading system                                |
| sqft_above    | Square footage of house apart from basement                                                                 |
| sqft_basement | Square footage of the basement                                                                              |
| yr_built      | Built Year                                                                                                  |
| yr_renovated  | Year when house was renovated                                                                               |
| zipcode       | Zip code                                                                                                    |
| lat           | Latitude coordinate                                                                                         |
| long          | Longitude coordinate                                                                                        |
| sqft_living15 | Living room area in 2015(implies-- some renovations) This might or might not have affected the lotsize area |
| sqft_lot15    | LotSize area in 2015(implies-- some renovations)                                                            |


## **Import the required libraries**


In [1]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented.
%pip install pandas
%pip install matplotlib 
%pip install scikit-learn
%pip install seaborn


# Note: If your environment doesn't support "!mamba install", use "!pip install"


[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip



Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.Requirement already satisfied: seaborn in c:\users\rcustodio\appdata\local\programs\python\python312\lib\site-packages (0.13.1)




[notice] A new release of pip is available: 23.3.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [18]:
%pip install seaborn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.3.1 -> 23.3.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
# Surpress warnings:
def warn(*args, **kwargs):
    pass
import warnings
warnings.warn = warn

In [4]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler,PolynomialFeatures
from sklearn.linear_model import LinearRegression
%matplotlib inline

# Module 1: Importing Data Sets


Download the dataset by running the cell below.


In [5]:
import piplite
await piplite.install('seaborn')

ModuleNotFoundError: No module named 'piplite'

In [90]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

ModuleNotFoundError: No module named 'pyodide'

In [ ]:
filepath='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/FinalModule_Coursera/data/kc_house_data_NaN.csv'

Load the csv:


In [ ]:
df = pd.read_csv(filepath)

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [6]:
filepath='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/FinalModule_Coursera/data/kc_house_data_NaN.csv'
df = pd.read_csv(filepath, header=None)

We use the method <code>head</code> to display the first 5 columns of the dataframe.


In [7]:
df.head(2)

0           1                2         3         4          5   \
0  NaN          id             date     price  bedrooms  bathrooms   
1  0.0  7129300520  20141013T000000  221900.0       3.0        1.0   

            6         7       8           9   ...     12          13  \
0  sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above   
1         1180      5650     1.0           0  ...      7        1180   

              14        15            16       17       18        19  \
0  sqft_basement  yr_built  yr_renovated  zipcode      lat      long   
1              0      1955             0    98178  47.5112  -122.257   

              20          21  
0  sqft_living15  sqft_lot15  
1           1340        5650  

[2 rows x 22 columns]

In [8]:
new_column_names = {
    0: 'NaN',
    1: 'id',
    2: 'date',
    3: 'price',
    4: 'bedrooms',
    5: 'bathrooms',
    6: 'sqft_living',
    7: 'sqft_lot',
    8: 'floors',
    9: 'waterfront',
    10: 'view',
    11: 'condition',
    12: 'grade',
    13: 'sqft_above',
    14: 'sqft_basement',
    15: 'yr_built',
    16: 'yr_renovated',
    17: 'zipcode',
    18: 'lat',
    19: 'long',
    20: 'sqft_living15',
    21: 'sqft_lot15'}

df.rename(columns=new_column_names, inplace=True)

In [9]:
df.to_csv("FINAL_Project_House_Sales_in_King_Count_USA-20231003-1696291200.jupyterlite", index=False)

### Question 1

Display the data types of each column using the function dtypes. Take a screenshot of your code and output. You will need to submit the screenshot for the final project. 


In [10]:
#Enter Your Code, Execute and take the Screenshot
df.dtypes

NaN              float64
id                object
date              object
price             object
bedrooms          object
bathrooms         object
sqft_living       object
sqft_lot          object
floors            object
waterfront        object
view              object
condition         object
grade             object
sqft_above        object
sqft_basement     object
yr_built          object
yr_renovated      object
zipcode           object
lat               object
long              object
sqft_living15     object
sqft_lot15        object
dtype: object

We use the method describe to obtain a statistical summary of the dataframe.


In [101]:
df.describe()

id             date     price bedrooms bathrooms sqft_living  \
count       21614            21614     21614    21601     21604       21614   
unique      21437              373      4029       13        30        1039   
top     795000620  20140623T000000  450000.0      3.0       2.5        1300   
freq            3              142       172     9824      5380         138   

       sqft_lot floors waterfront   view  ...  grade sqft_above sqft_basement  \
count     21614  21614      21614  21614  ...  21614      21614         21614   
unique     9783      7          3      6  ...     13        947           307   
top        5000    1.0          0      0  ...      7       1300             0   
freq        358  10680      21450  19489  ...   8981        212         13126   

       yr_built yr_renovated zipcode      lat     long sqft_living15  \
count     21614        21614   21614    21614    21614         21614   
unique      117           71      71     5035      753           778   
top        2014            0   98103  47.6846  -122.29          1540   
freq        559        20699     602       17      116           197   

       sqft_lot15  
count       21614  
unique       8690  
top          5000  
freq          427  

[4 rows x 21 columns]

# Module 2: Data Wrangling


### Question 2

Drop the columns <code>"id"</code>  and <code>"Unnamed: 0"</code> from axis 1 using the method <code>drop()</code>, then use the method <code>describe()</code> to obtain a statistical summary of the data. Make sure the <code>inplace</code> parameter is set to <code>True</code>. Take a screenshot of your code and output. You will need to submit the screenshot for the final project. 


In [100]:
#Enter Your Code, Execute and take the Screenshot
df.drop(columns=['NaN'], axis=1, inplace=True)

We can see we have missing values for the columns <code> bedrooms</code>  and <code> bathrooms </code>


In [71]:
df.describe()

id             date     price bedrooms bathrooms sqft_living  \
count       21614            21614     21614    21601     21604       21614   
unique      21437              373      4029       13        30        1039   
top     795000620  20140623T000000  450000.0      3.0       2.5        1300   
freq            3              142       172     9824      5380         138   

       sqft_lot floors waterfront   view  ...  grade sqft_above sqft_basement  \
count     21614  21614      21614  21614  ...  21614      21614         21614   
unique     9783      7          3      6  ...     13        947           307   
top        5000    1.0          0      0  ...      7       1300             0   
freq        358  10680      21450  19489  ...   8981        212         13126   

       yr_built yr_renovated zipcode      lat     long sqft_living15  \
count     21614        21614   21614    21614    21614         21614   
unique      117           71      71     5035      753           778   
top        2014            0   98103  47.6846  -122.29          1540   
freq        559        20699     602       17      116           197   

       sqft_lot15  
count       21614  
unique       8690  
top          5000  
freq          427  

[4 rows x 21 columns]

In [72]:
missing_data = df.isnull()
print(missing_data.head())
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

      id   date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  False  False  False     False      False        False     False   False   
1  False  False  False     False      False        False     False   False   
2  False  False  False     False      False        False     False   False   
3  False  False  False     False      False        False     False   False   
4  False  False  False     False      False        False     False   False   

   waterfront   view  ...  grade  sqft_above  sqft_basement  yr_built  \
0       False  False  ...  False       False          False     False   
1       False  False  ...  False       False          False     False   
2       False  False  ...  False       False          False     False   
3       False  False  ...  False       False          False     False   
4       False  False  ...  False       False          False     False   

   yr_renovated  zipcode    lat   long  sqft_living15  sqft_lot15  
0         False    False

In [102]:
print("number of NaN values for the column bedrooms :", df['bedrooms'].isnull().sum())
print("number of NaN values for the column bathrooms :", df['bathrooms'].isnull().sum())

number of NaN values for the column bedrooms : 13
number of NaN values for the column bathrooms : 10


We can replace the missing values of the column <code>'bedrooms'</code> with the mean of the column  <code>'bedrooms' </code> using the method <code>replace()</code>. Don't forget to set the <code>inplace</code> parameter to <code>True</code>


In [74]:
df['bedrooms'] = pd.to_numeric(df['bedrooms'], errors='coerce')

In [48]:
#Primero se debe convertir los INT to FLOAT

In [75]:
mean = df['bedrooms'].mean()
df['bedrooms'].replace(np.nan,mean, inplace=True)

We also replace the missing values of the column <code>'bathrooms'</code> with the mean of the column  <code>'bathrooms' </code> using the method <code>replace()</code>. Don't forget to set the <code> inplace </code>  parameter top <code> True </code>


In [76]:
df['bathrooms'] = pd.to_numeric(df['bathrooms'], errors='coerce')

In [77]:
mean=df['bathrooms'].mean()
df['bathrooms'].replace(np.nan,mean, inplace=True)

In [78]:
print("number of NaN values for the column bedrooms :", df['bedrooms'].isnull().sum())
print("number of NaN values for the column bathrooms :", df['bathrooms'].isnull().sum())

number of NaN values for the column bedrooms : 0
number of NaN values for the column bathrooms : 0


In [79]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0          id             date     price   3.37287   2.115736  sqft_living   
1  7129300520  20141013T000000  221900.0   3.00000   1.000000         1180   
2  6414100192  20141209T000000  538000.0   3.00000   2.250000         2570   
3  5631500400  20150225T000000  180000.0   2.00000   1.000000          770   
4  2487200875  20141209T000000  604000.0   4.00000   3.000000         1960   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement   
1      5650     1.0           0     0  ...      7        1180              0   
2      7242     2.0           0     0  ...      7        2170            400   
3     10000     1.0           0     0  ...      6         770              0   
4      5000     1.0           0     0  ...      7        1050            910   

   yr_built  yr_renovated  zipcode                 lat                 long  \
0  yr_built  yr_renovated  zipcode                 lat                 long   
1      1955             0    98178             47.5112             -122.257   
2      1951          1991    98125  47.721000000000004             -122.319   
3      1933             0    98028             47.7379  -122.23299999999999   
4      1965             0    98136             47.5208  -122.39299999999999   

   sqft_living15  sqft_lot15  
0  sqft_living15  sqft_lot15  
1           1340        5650  
2           1690        7639  
3           2720        8062  
4           1360        5000  

[5 rows x 21 columns]

In [80]:
df.to_csv("House_sales_in_King_Count.csv", index=False)

In [1]:
unique_characters = df['id'].str.cat(sep='').unique()
unique_characters_list = list(unique_characters)
print(unique_characters_list)

NameError: name 'df' is not defined

In [84]:
#The error you're encountering, "AttributeError: 'str' object has no attribute 'unique'," occurs because 
#you are trying to apply the unique() method to a string object (unique_characters) rather than to a Pandas Series or DataFrame.
#If your goal is to find unique characters within the concatenated string of the 
#'id' column, you can use Python's built-in set() function to achieve this. Here's the modified code:

In [82]:
unique_characters = df['id'].str.cat(sep='')
unique_characters_set = set(unique_characters)
unique_characters_list = list(unique_characters_set)
print(unique_characters_list)


['d', '0', '9', '2', '5', '6', '4', '3', '8', 'i', '7', '1']


# Module 3: Exploratory Data Analysis


### Question 3

Use the method <code>value_counts</code> to count the number of houses with unique floor values, use the method <code>.to_frame()</code> to convert it to a data frame. Take a screenshot of your code and output. You will need to submit the screenshot for the final project. 


In [ ]:
#Enter Your Code, Execute and take the Screenshot


### Question 4

Use the function <code>boxplot</code> in the seaborn library  to  determine whether houses with a waterfront view or without a waterfront view have more price outliers. Take a screenshot of your code and boxplot. You will need to submit the screenshot for the final project. 


### Question 5

Use the function <code>regplot</code>  in the seaborn library  to  determine if the feature <code>sqft_above</code> is negatively or positively correlated with price. Take a screenshot of your code and scatterplot. You will need to submit the screenshot for the final project. 


In [ ]:
#Enter Your Code, Execute and take the Screenshot

We can use the Pandas method <code>corr()</code>  to find the feature other than price that is most correlated with price.


In [1]:
df.corr()['price'].sort_values()

NameError: name 'df' is not defined

# Module 4: Model Development


We can Fit a linear regression model using the  longitude feature <code>'long'</code> and  caculate the R^2.


In [ ]:
X = df[['long']]
Y = df['price']
lm = LinearRegression()
lm.fit(X,Y)
lm.score(X, Y)

### Question  6

Fit a linear regression model to predict the <code>'price'</code> using the feature <code>'sqft_living'</code> then calculate the R^2. Take a screenshot of your code and the value of the R^2. You will need to submit it for the final project.


In [ ]:
#Enter Your Code, Execute and take the Screenshot

### Question 7

Fit a linear regression model to predict the <code>'price'</code> using the list of features:


In [ ]:
features =["floors", "waterfront","lat" ,"bedrooms" ,"sqft_basement" ,"view" ,"bathrooms","sqft_living15","sqft_above","grade","sqft_living"]     

Then calculate the R^2. Take a screenshot of your code and the value of the R^2. You will need to submit it for the final project.


In [ ]:
#Enter Your Code, Execute and take the Screenshot

### This will help with Question 8

Create a list of tuples, the first element in the tuple contains the name of the estimator:

<code>'scale'</code>

<code>'polynomial'</code>

<code>'model'</code>

The second element in the tuple  contains the model constructor

<code>StandardScaler()</code>

<code>PolynomialFeatures(include_bias=False)</code>

<code>LinearRegression()</code>


In [ ]:
Input=[('scale',StandardScaler()),('polynomial', PolynomialFeatures(include_bias=False)),('model',LinearRegression())]

### Question 8

Use the list to create a pipeline object to predict the 'price', fit the object using the features in the list <code>features</code>, and calculate the R^2. Take a screenshot of your code and the value of the R^2. You will need to submit it for the final project.


In [ ]:
#Enter Your Code, Execute and take the Screenshot

# Module 5: Model Evaluation and Refinement


Import the necessary modules:


In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
print("done")

We will split the data into training and testing sets:


In [ ]:
features =["floors", "waterfront","lat" ,"bedrooms" ,"sqft_basement" ,"view" ,"bathrooms","sqft_living15","sqft_above","grade","sqft_living"]    
X = df[features]
Y = df['price']

x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.15, random_state=1)


print("number of test samples:", x_test.shape[0])
print("number of training samples:",x_train.shape[0])

### Question 9

Create and fit a Ridge regression object using the training data, set the regularization parameter to 0.1, and calculate the R^2 using the test data. Take a screenshot of your code and the value of the R^2. You will need to submit it for the final project.


In [ ]:
from sklearn.linear_model import Ridge

In [ ]:
#Enter Your Code, Execute and take the Screenshot

### Question 10

Perform a second order polynomial transform on both the training data and testing data. Create and fit a Ridge regression object using the training data, set the regularisation parameter to 0.1, and calculate the R^2 utilising the test data provided. Take a screenshot of your code and the R^2. You will need to submit it for the final project.


In [ ]:
#Enter Your Code, Execute and take the Screenshot

<p>Once you complete your notebook you will have to share it. You can download the notebook by navigating to "File" and clicking on "Download" button.
        <p><img width="600" src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%206/images/DA0101EN_FA_Image21.png" alt="share notebook" style="display: block; margin-left: auto; margin-right: auto;"></p>
        <p></p>
<p>This will save the (.ipynb) file on your computer. Once saved, you can upload this file in the "My Submission" tab, of the "Peer-graded Assignment" section.  
          


<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.


Other contributors: <a href="https://www.linkedin.com/in/michelleccarey/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2022-01-01">Michelle Carey</a>, <a href="https://www.linkedin.com/in/jiahui-mavis-zhou-a4537814a?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2022-01-01">Mavis Zhou</a>


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By      | Change Description                           |
| ----------------- | ------- | --------------- | -------------------------------------------- |
| 2020-12-01        | 2.2     | Aije Egwaikhide | Coverted Data describtion from text to table |
| 2020-10-06        | 2.1     | Lakshmi Holla   | Changed markdown instruction of Question1    |
| 2020-08-27        | 2.0     | Malika Singla   | Added lab to GitLab                          |
| 2022-06-13        | 2.3     | Svitlana Kramar | Updated Notebook sharing instructions        |
| <hr>              |         |                 |                                              |

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
